Migrate CI to GitHub Actions (v2)

giovana-morais opened the pull request:
based on #217 (unfortunately, i can't add commits to his branch)

## changelog
* remove Travis files
* remove requirements.txt
* add GitHub Action ci.yml file
* bump minimal mir_eval version to 0.8.2

bmcfee commented on setup.py:
```diff
@@ -39,7 +39,7 @@ def load_source(modname, filename):
         "Programming Language :: Python :: 3.13",
     ],
-    python_requires=">=3.9",
+    python_requires=">=3.10",
```
3.9 is still live, we shouldn't drop it yet.

giovana-morais replied:
my bad, i misread the comment in the other thread. i'll work on the namespace issue then!

giovana-morais commented on tests/test_sonify.py:
```diff
@@ -24,6 +24,7 @@ def test_bad_sonify():
         jams.sonify.sonify(ann)
 
 
+@pytest.mark.xfail
```
temporary until mir_eval 0.8.1. should i leave a comment?

bmcfee replied:
nm just saw the note in the pr :)

bmcfee commented:
mir_eval 0.8.1 is on pypi now; should hit conda-forge by this afternoon.

giovana-morais commented:
> mir_eval 0.8.1 is on pypi now; should hit conda-forge by this afternoon.

i ran the sonify tests and some are failing but for a different reason now. i'll mark those as xfail again and i can investigate and provide more details afterward in the `mir_eval` repo if needed.

```
FAILED tests/test_sonify.py::test_note_hz - AttributeError: 'list' object has no attribute 'shape'
FAILED tests/test_sonify.py::test_note_hz_nolength - AttributeError: 'list' object has no attribute 'shape'
FAILED tests/test_sonify.py::test_note_midi - AttributeError: 'list' object has no attribute 'shape'
```

other than this, i added back the support to python 3.9 by using `pkg_resource`. i didn't figure out what was the deal with `importlib.resources.files`, so i just moved forward.

giovana-morais commented on jams/__init__.py:
```diff
@@ -19,11 +20,17 @@
-for ns in chain(*map(lambda p: p.rglob('*.json'), resources.files('jams.schemata.namespaces').iterdir())):
-    schema.add_namespace(ns)
+if sys.version_info < (3, 10):
+    from pkg_resources import resource_filename
```
thought that leaving the import statement here might be better than at the beginning of the code to avoid deprecation warning for newer python versions?

bmcfee replied:
Yeah that should be fine.

bmcfee commented on tests/test_sonify.py:
```diff
@@ -38,6 +38,7 @@ def test_duration(ns, sr, duration):
         assert len(y) == int(sr * duration)
 
 
+@pytest.mark.xfail
```
I think these tests are failing because of the following code:

https://github.com/marl/jams/blob/58a24d5513749bd4fd64c66d7341fce615f93208/jams/sonify.py#L161-L173

The `annotation.to_interval_values()` function is returning lists which need to be cast to np.ndarrays for sonification, as is done elsehwere in `jams.sonify`.

If you replace the following line:
https://github.com/marl/jams/blob/58a24d5513749bd4fd64c66d7341fce615f93208/jams/sonify.py#L171-L173
by
```python
    return filter_kwargs(mir_eval.sonify.time_frequency,
                         gram, np.asarray(pitches), np.asarray(intervals),
                         sr, length=length, **kwargs)
```

it will fix a couple of the errors.  There does appear to be a separate error with the `nolength` test below that will require some patching in mir_eval.  I'll look into it.

bmcfee commented on tests/test_sonify.py:
```diff
@@ -46,6 +47,7 @@ def test_note_hz():
     assert len(y) == 8000 * 2
 
 
+@pytest.mark.xfail
```
This one is failing due to a bug in the `_const_interpolator` mode of mir_eval.sonify :facepalm: I'll fix it and get a 0.8.2 release going shortly...

bmcfee replied:
Fix is in https://github.com/mir-evaluation/mir_eval/pull/418

giovana-morais replied:
should i wait for the release and then test again as the last sanity check?

bmcfee replied:
0.8.2 is up on pypi now; conda-forge is building and should be up within the hour.

giovana-morais replied:
cool, everything works now! resolving this.

bmcfee commented on setup.py:
```diff
@@ -49,7 +49,7 @@ def load_source(modname, filename):
         'six',
         'decorator',
-        'mir_eval>0.7',
+        'mir_eval>=0.8'
```
```suggestion
        'mir_eval>=0.8.2'
```

bmcfee commented on .github/workflows/jams.yml:
```diff
@@ -0,0 +1,30 @@
+#  push:
+#    branches: [ master ]
+#  pull_request:
+#    branches: [ master ]
```
We should also have these enabled for PR

giovana-morais replied:
should i remove the push one?

bmcfee replied:
We should have both, see https://github.com/librosa/librosa/blob/894942673d55aa2206df1296b6c4c50827c7f1d6/.github/workflows/ci.yml#L3-L9

giovana-morais replied:
interesting, by adding both i can't really run the actions on my fork. i'm assuming that it's because i'm not in the main branch?

bmcfee replied:
Right; on your fork it's not a PR against master, and also not a push to master.  But once we merge this so the CI is live on the main repo, it should work again.

giovana-morais replied:
understood, thanks :smile:

bmcfee commented on .github/workflows/jams.yml:
```diff
@@ -0,0 +1,30 @@
+name: JSON Annotated Music Specification for Reproducible MIR Research
```
Prefer to rename this file to `test.yml` or `ci.yml` - we'll likely want to add other workflows, and "jams" is not an informative name

bmcfee commented:
A couple more comments, but this otherwise is looking good!
